frontend/src/components/widgets/DataFrame/hooks/useDataLoader.ts

```diff
@@ -0,0 +1,318 @@
+ *
+ * @param columnProps - The column properties to apply the config to.
+ * @param columnConfigMapping - The user-defined column configuration mapping.
+ *
+ * @return the column properties with the config applied.
+ */
+export function applyColumnConfig(
+  columnProps: BaseColumnProps,
+  columnConfigMapping: Map<string | number, ColumnConfigProps>
+): BaseColumnProps {
+  if (!columnConfigMapping) {
+    // No column config configured
+    return columnProps
+  }
+
+  let columnConfig
+  if (columnConfigMapping.has(columnProps.title)) {
+    // Config is configured based on the column title
+    columnConfig = columnConfigMapping.get(columnProps.title)
+  } else if (
+    columnConfigMapping.has(
+      `${COLUMN_POSITION_PREFIX}${columnProps.indexNumber}`
+    )
+  ) {
+    // Config is configured based on the column position, e.g. col:0 -> first column
+    columnConfig = columnConfigMapping.get(
+      `${COLUMN_POSITION_PREFIX}${columnProps.indexNumber}`
+    )
+  } else if (
+    columnProps.isIndex &&
+    columnConfigMapping.has(INDEX_IDENTIFIER)
+  ) {
+    // Config is configured for the index column (or all index columns for multi-index)
+    columnConfig = columnConfigMapping.get(INDEX_IDENTIFIER)
+  }
+
+  if (!columnConfig) {
+    // No column config found for this column
+    return columnProps
+  }
+
+  return {
+    ...columnProps,
+    // Update title:
+    ...(notNullOrUndefined(columnConfig.title)
+      ? {
+          title: columnConfig.title,
+        }
+      : {}),
+    // Update width:
+    ...(notNullOrUndefined(columnConfig.width)
+      ? {
+          width: columnConfig.width,
+        }
+      : {}),
+    // Select a column type:
+    ...(notNullOrUndefined(columnConfig.type)
+      ? {
+          customType: columnConfig.type.toLowerCase().trim(),
+        }
+      : {}),
+    // Update editable state:
+    ...(notNullOrUndefined(columnConfig.editable)
+      ? {
+          isEditable: columnConfig.editable,
+        }
+      : {}),
+    // Update hidden state:
+    ...(notNullOrUndefined(columnConfig.hidden)
+      ? {
+          isHidden: columnConfig.hidden,
+        }
+      : {}),
+    // Add column type metadata:
+    ...(notNullOrUndefined(columnConfig.metadata)
+      ? {
+          columnTypeMetadata: columnConfig.metadata,
+        }
+      : {}),
+    // Add column alignment:
+    ...(notNullOrUndefined(columnConfig.alignment)
+      ? {
+          contentAlignment: columnConfig.alignment,
+        }
+      : {}),
+  } as BaseColumnProps
+}
+
+/**
+ * Extracts the user-defined column configuration from the proto message.
+ *
+ * @param element - The proto message of the dataframe element.
+ *
+ * @returns the user-defined column configuration.
+ */
+export function getColumnConfig(element: ArrowProto): Map<string, any> {
+  if (!element.columns) {
+    return new Map()
+  }
+  try {
+    return new Map(Object.entries(JSON.parse(element.columns)))
```

Comment:
This just a lazy action?

Reply:
What do you mean with lazy action?

The column config is currently not put into the state and this will be parsed on every render. It most likely could be put into the state to get some optimization. This does not seem to have a performance impact, but I have this on my fast-follow list to evaluate putting it into the state.